nexus/db-queries/src/db/queries/sled_reservation.rs

```diff
@@ -0,0 +1,385 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Implementation of queries for affinity groups
+
+use crate::db::model::AffinityPolicyEnum;
+use crate::db::model::Resources;
+use crate::db::model::SledResourceVmm;
+use crate::db::raw_query_builder::QueryBuilder;
+use crate::db::raw_query_builder::TypedSqlQuery;
+use diesel::sql_types;
+use omicron_uuid_kinds::GenericUuid;
+use omicron_uuid_kinds::InstanceUuid;
+
+/// Return all possible Sleds where we might perform allocation
+///
+/// The rows returned by this CTE indicate:
+///
+/// - The Sled which we're considering
+/// - A bool indicating whether the allocation fits
+/// - Affinity Policy
+/// - Anti-Affinity Policy
+///
+/// Generally, we'd like to only return sleds where an allocation of [Resources]
+/// could fit. However, we also need to observe all affinity groups, in case
+/// affinity group policy forces us to allocate on a sled where insufficient
+/// space exists.
+///
+/// Note that we don't bother checking if the VMM has already been provisioned,
+/// we're just searching for spots where it might fit.
+pub fn sled_find_targets_query(
+    instance_id: InstanceUuid,
+    resources: &Resources,
+) -> TypedSqlQuery<(
+    sql_types::Uuid,
+    sql_types::Bool,
+    sql_types::Nullable<AffinityPolicyEnum>,
+    sql_types::Nullable<AffinityPolicyEnum>,
+)> {
+    QueryBuilder::new().sql("
+        WITH sled_targets AS (
+            SELECT sled.id as sled_id
+            FROM sled
+            LEFT JOIN sled_resource_vmm
+            ON sled_resource_vmm.sled_id = sled.id
+            WHERE
+                sled.time_deleted IS NULL AND
+                sled.sled_policy = 'in_service' AND
+                sled.sled_state = 'active'
+            GROUP BY sled.id
+            HAVING
+                COALESCE(SUM(CAST(sled_resource_vmm.hardware_threads AS INT8)), 0) + "
+            ).param().sql(" <= sled.usable_hardware_threads AND
+                COALESCE(SUM(CAST(sled_resource_vmm.rss_ram AS INT8)), 0) + "
+            ).param().sql(" <= sled.usable_physical_ram AND
+                COALESCE(SUM(CAST(sled_resource_vmm.reservoir_ram AS INT8)), 0) + "
+            ).param().sql(" <= sled.reservoir_size
+        ),
+        our_aa_groups AS (
+            SELECT group_id
+            FROM anti_affinity_group_instance_membership
+            WHERE instance_id = ").param().sql("
+        ),
+        other_aa_instances AS (
+            SELECT anti_affinity_group_instance_membership.group_id,instance_id
+            FROM anti_affinity_group_instance_membership
+            JOIN our_aa_groups
+            ON anti_affinity_group_instance_membership.group_id = our_aa_groups.group_id
+            WHERE instance_id != ").param().sql("
+        ),
```

Comment:
Optional: I'd almost be inclined to make these individual subqueries into individual `const &str`s, in part so they can be shared between `sled_insert_resource_query` and `sled_find_targets_query`, and in part so you can add doc comments to them (or inline comments in the queries) to help describe what's going on. Up to you how far to go here, though.

Reply:
This gets a fair bit more complicated in https://github.com/oxidecomputer/omicron/pull/7572 - do you mind if I punt some of the cleanup to that PR, to avoid a merge conflict with myself?

Reply:
Go for it!